### PR TITLE
Fix PSATD equations used with PML

### LINKED ISSUE #1420: Boundary conditions with PSATD
I've been testing the boundary conditions in a very simple 2D case with a laser aligned with one of the axes, either with `do_pml = 0` or `do_pml = 1`. When I use the Yee solver, the behaviour is good: `do_pml = 0` corresponds to reflecting boundary conditions (although I don't think this is explicitly stated in the docs I'm assuming this is the expected behaviour) and `do_pml = 1` corresponds to absorbing boundary conditions.

Now if I do the same test with the PSATD solver, I find that in both cases the boundary conditions are not working as well, see image below.

Boundary_Yee_vs_PSATD

Do you know what is happening? (for `do_pml = 0` I was thinking that maybe the fields are never set to 0 in the guard cells but in this case I would expect something similar to periodic conditions so I don't really know)

The input file I used is the following:

```
#################################
####### GENERAL PARAMETERS ######
#################################
max_step = 3800
amr.n_cell =  2048 3072
warpx.numprocs = 1 1
amr.max_level = 0
geometry.coord_sys   = 0                  # 0: Cartesian
geometry.is_periodic = 0 0 0     # Is periodic?
geometry.prob_lo     = -7e-6  -16.0e-6     # physical domain
geometry.prob_hi     =  7e-6   8.0e-6

#################################
########## DIAGNOSTICS ##########
#################################
diagnostics.diags_names = diag0
diag0.period = 3800
diag0.diag_type = Full
diag0.fields_to_plot = By

#################################
############ NUMERICS ###########
#################################
interpolation.nox = 3
interpolation.noy = 3
interpolation.noz = 3
warpx.cfl = 0.99
warpx.do_pml = 1

#################################
############# LASER #############
#################################
lasers.names        = laser
laser.position     = 0. 0. -10e-6 # This point is on the laser plane
laser.direction    = 0. 0. 1.     # The plane normal direction
laser.polarization = 1. 0. 0.     # The main polarization vector
laser.e_max        =  1.5e+17
laser.wavelength = 0.8e-6         # The wavelength of the laser (in meters)
laser.profile      = Gaussian
laser.profile_waist = 2.e-6
laser.profile_duration = 6.e-15
laser.profile_t_peak = 20.e-15
laser.profile_focal_distance = 6.e-6
```

The only things I've changed are:
- The compilations options (`make -j32 USE_GPU=TRUE DIM=2` or `make -j32 USE_GPU=TRUE DIM=2 USE_PSATD=TRUE`)
- `warpx.do_pml` which is either 1 or 0
- The final timestep, which is also the timestep of diagnostics. It is 3800 with Yee and 2860 with PSATD (due to different values for the temporal step).

Note that this test has been done on GPU, I'll run a new test to see if the behaviour changes on CPU.
Edit: done the test on CPU, there's no obvious difference between GPU and CPU.

### PR DESCRIPTION
**Pull request by EZoni**
**Goal**
Fix the PSATD equations used with PML and close #1420.

**Description**
The new PSATD equations require 22 distinct coefficients. The implementation did not change the design of the relevant classes.

**Tests**
The PR was tested separately, using the setup of the CI test `pml_x_psatd`:

<p align="center">
    
</p>

The benchmark for the CI test `pml_x_psatd` was updated. Also the new analysis simply checks whether the numerical reflectivity is below a certain threshold (`1e-6`). I also removed a check on charge conservation with very large tolerance (`2e-2`), since this test does not run with current correction.

### CODE REVIEW

**Review by NeilZaim — APPROVED** [review]
Thanks a lot for the PR! To me the code is looking good (I'll trust you with the computation of the coefficients and with the field update :D ), I just have a minor suggestion regarding the test.

File: Examples/Tests/PML/analysis_pml_psatd.py [hunk 1]
```diff
@@ -20,7 +19,7 @@
 ############################
 ### INITIAL LASER ENERGY ###
 ############################
-energy_start = 9.1301289517e-08
+energy_start = 7.297301362346945e-08 # electromagnetic energy at iteration 50
```
Comment (NeilZaim) [inline Examples/Tests/PML/analysis_pml_psatd.py:22]: Maybe we could also compute the starting energy from the diags and assert that the value is not too different from the one here?

This would pretty much be another checksum, but in case the checksums are failing this one would indicate if the difference is coming from the antenna+solver or from the PML.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp [hunk 2]
```diff
@@ -124,32 +223,87 @@ void PMLPsatdAlgorithm::InitializeSpectralCoefficients (
...
+            constexpr Real c2 = c*c;
+
+            // Coefficients for k_norm = 0 do not need to be set
+            if (k_norm != 0.) {
+
+                const Real omega = c * k_norm;
+
+                C(i,j,k)    = std::cos(omega * dt);
+                S_ck(i,j,k) = std::sin(omega * dt) / omega;
+
```
Comment (RemiLehe) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp]: It seems that the coefficients `C1` to `C22` can all be computed with simple operation from a small number of coefficients (e.g. `C`, `S_ck`, etc.). Therefore, I would suggest to only save `C` and `S_ck` (and maybe `inv_k2 = 1./k2`) in the class instance, and compute `C1` to `C22` on the fly, inside the field pusher kernel.

This would save memory, esp. on GPU, since these 22 3D arrays might take up quite a bit of space.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp [hunk 3]
```diff
@@ -82,25 +127,76 @@ PMLPsatdAlgorithm::pushSpectralFields(SpectralFieldData& f) const{
...
-            fields(i,j,k,Idx::Byx) = C*fields(i,j,k,Idx::Byx) + S_ck*I*kx*Ez_old;
-            fields(i,j,k,Idx::Bzx) = C*fields(i,j,k,Idx::Bzx) - S_ck*I*kx*Ey_old;
-            fields(i,j,k,Idx::Bzy) = C*fields(i,j,k,Idx::Bzy) + S_ck*I*ky*Ex_old;
+
+            const Real k_norm = std::sqrt(kx*kx + ky*ky + kz*kz);
+
+            constexpr Real c2 = PhysConst::c * PhysConst::c;
+
+            const Complex I = Complex{0._rt, 1._rt};
+
```
Comment (RemiLehe) [inline Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp:96]: Could we calculate the coefficients `C1` to `C22` on the fly here? (See comment in the function `InitializeSpectralCoefficients`)

**Review by RemiLehe — CHANGES_REQUESTED** [review]

**Comment by RemiLehe** [comment]
Nice! Thanks a lot for this important change to the PML equations!! 
I have a comment on the coefficients `C1` to `C22` (see inline comments), but otherwise the PR looks good to me.

File: Examples/Tests/PML/analysis_pml_psatd.py (hunk 1 above)
Comment (EZoni) [reply Examples/Tests/PML/analysis_pml_psatd.py:22]: @NeilZaim Thank you for your suggestion. I implemented the change in https://github.com/ECP-WarpX/WarpX/pull/1513/commits/e231246048ea92b482ff463b8154384f45df9f83, I hope this is what you had in mind.

File: Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp (hunk 2 above)
Comment (EZoni) [reply Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PMLPsatdAlgorithm.cpp]: @RemiLehe Thank you for your suggestion. I implemented the change in https://github.com/ECP-WarpX/WarpX/pull/1513/commits/358e7242a4c09db138f0873490397858191ddbaf.

**Review by RemiLehe — APPROVED** [review]
